src/geometry/shamos_hoey.rs

```diff
@@ -0,0 +1,309 @@
+/// Shamos-Hoey Algorithm. Given a set of line segments, finds if any two line segments intersect.
+/// Some implementations assume that there are no vertical lines. This implementation supports
+/// vertical lines.
+use super::point::Point;
+use super::segment::Segment;
+use std::cmp::{Ord, Ordering};
+use std::collections::BTreeSet;
+
+enum PointType {
+    Start,
+    End,
+}
+
+type SegmentAndPoint<'a> = (&'a Segment, &'a Point, PointType);
+struct SegmentWrapper<'a> {
+    segment: &'a Segment,
+    // `x` holds the x value when the segment was added to the set. We order segments at point `x`
+    // by their y-value at `x`. Note that if the order of two segments ever changes, it means that
+    // they intersected, which our algorithm can detect.
+    //
+    // Ideally we could track a single value of x as it changes through the algorithm iterations.
+    // We need to implement `Ord` for the struct so that we could insert it into the RB-tree. The
+    // Ord implementation would not have access to a `x` local variable in `shamos_hoey()`. We keep
+    // keep track of x in the struct as a workaround. When comparing two segments, we take the max
+    // of the two segments' x's, which represents the most up to date `x`.
+    x: f64,
+    // `point` is used soley for comparisons involving a vertical segment. For most comparisons, we
+    // compute the y-value at a given x, but vertical lines would have multiple y-values. For
+    // vertical lines, we order them by the endpoint from bottom to top.
+    point: &'a Point,
+}
+
+impl<'a> SegmentWrapper<'a> {
+    fn new(segment: &'a Segment, x: f64, point: &'a Point) -> SegmentWrapper<'a> {
+        SegmentWrapper { segment, x, point }
+    }
+}
+
+impl Segment {
+    // Returns (start, end) points
+    fn get_start_end(&self) -> (&Point, &Point) {
+        match point_cmp(&self.a, &self.b) {
+            Ordering::Less | Ordering::Equal => (&self.a, &self.b),
+            Ordering::Greater => (&self.b, &self.a),
+        }
+    }
+}
+
+impl<'a> Ord for SegmentWrapper<'a> {
+    fn cmp(&self, other: &SegmentWrapper) -> Ordering {
+        // For vertical segments, we compare using the endpoint.
+        let x = if self.x < other.x { other.x } else { self.x };
```

Comment:
In the algorithm, we should be using the latest value of x in the left-to-right sweep. However, the `Ord` implementation does not have access to the local variable x in `shamos_hoey()`,but we need to implement `Ord` if we want to use the stdlib btreeset implementation. The balanced tree implementations in this repo also compare elements using the `Ord` trait. ` I would strongly prefer to avoid using a ad-hoc balanced tree implementation. We get around this issue by storing x in the SegmentWrapper, (see the comment in the struct above).

I was hoping that I can implement `Ord` using a lambda that captures the local variable x, but I don't think there is a way of doing that in Rust. I also tried storing x into a global variable, but that produces inconsistent results among test threads.

Are there Rust experts who can help with making this code a little cleaner?

Reply:
You might also want to write about it here https://discordapp.com/channels/808045925556682782/1022517736032714822. It's more likely to be discussed in Discord

Reply:
Thank you. Dropped a post in https://discord.com/channels/808045925556682782/1022517736032714822/threads/1131472585390047232